google/cloud/spanner/doc/spanner-main.dox

```diff
@@ -0,0 +1,220 @@
+
+In your [WORKSPACE][workspace-definition-link] file add a dependency to download
+and install the library, for example:
+
+[workspace-definition-link]: https://docs.bazel.build/versions/master/build-ref.html#packages_targets
+
+@code {.py}
+# Change the version and SHA256 hash as needed.
+http_archive(
+    name = "com_github_googleapis_google_cloud_cpp_spanner",
+    url = "https://github.com/googleapis/google-cloud-cpp-spanner/archive/e2e73b6b044233a51d00090f1f09cd1b7b3af76b.tar.gz",
+    strip_prefix = "google-cloud-cpp-spanner-e2e73b6b044233a51d00090f1f09cd1b7b3af76b",
+    sha256 = "f1b4f3a35b89545d904a9484f38b984cd21f1d8c2e50f9ab21f3ba988cfe11e2",
+)
+@endcode
+
+Then load the dependencies of the library:
+
+```{.py}
+load("@com_github_googleapis_google_cloud_cpp_spanner//bazel:google_cloud_cpp_spanner_deps.bzl", "google_cloud_cpp_spanner_deps")
+
+google_cloud_cpp_spanner_deps()
+
+# Configure @com_google_googleapis to only compile C++ and gRPC:
+load("@com_google_googleapis//:repository_rules.bzl", "switched_rules_by_language")
+
+switched_rules_by_language(
+    name = "com_google_googleapis_imports",
+    cc = True,
+    grpc = True,
+)
+
+# Have to manually call the corresponding function for google-cloud-cpp and
+# gRPC:  https://github.com/bazelbuild/bazel/issues/1550
+load("@com_github_googleapis_google_cloud_cpp//bazel:google_cloud_cpp_deps.bzl", "google_cloud_cpp_deps")
+
+google_cloud_cpp_deps()
+
+load("@com_github_grpc_grpc//bazel:grpc_deps.bzl", "grpc_deps")
+
+grpc_deps()
+```
+
+You can now use the library as a dependency in your BUILD files, for example:
+
+```{.py}
+cc_binary(
+    name = "my_program",
+    srcs = [
+        "my_program.cc",
+    ],
+    deps = [
+        "@com_github_googleapis_google_cloud_cpp_spanner//google/cloud/spanner:spanner_client",
+    ],
+)
+```
+
+#### Integrating with Make
+
+Installing the client library with CMake also creates `pkg-config` support files
+for application developers that prefer to use `Make` as their build system. Once
+the library is installed, you can use it in your `Makefile` like any other
+`pkg-config` module:
+
+```
+# Configuration variables to compile and link against the Cloud Spanner C++
+# client library.
+SPANNER_CXXFLAGS   := $(shell pkg-config spanner_client --cflags)
+SPANNER_CXXLDFLAGS := $(shell pkg-config spanner_client --libs-only-L)
+SPANNER_LIBS       := $(shell pkg-config spanner_client --libs-only-l)
+
+# A target using the Cloud Spanner C++ client library.
+my_program: my_program.cc
+    $(CXX) $(CXXFLAGS) $(SPANNER_CXXFLAGS) $(SPANNER_CXXLDFLAGS) -o $@ $^ $(SPANNER_LIBS)
+```
+
+### Error Handling
+
+[status-or-header]: https://github.com/googleapis/google-cloud-cpp/blob/master/google/cloud/status_or.h
+
+This library never throws exceptions to signal error. In general, the library
+returns a [`StatusOr<T>`][status-or-header] if an error is possible. Some
+functions return objects that already have an existing error handling mechanism,
```

Comment:
Does the Spanner library actually return any type that signals errors using something other than StatusOr? I don't think returns any ostreams. It might grow these things, but I don't think we have them yet.

Do we want to keep this blurb in case we grow these new error-signaling types, or remove it and add it back if/when we need to?

Honestly, I don't have a strong opinion here. Mostly just confirming if my understanding of the library's current state is accruate.

Reply:
> Does the Spanner library actually return any type that signals errors using something other than StatusOr?

We return `future<StatusOr<T>>`. We are also planning to return `ResultSet` both of which use `StatusOr<>` to signal an error "later", but are not wrapped by a `StatusOr<>`. The stuff about iostream was (obviously) a cut&paste error.